### PR TITLE
[allwinner] add $DEVICES support and UBOOT_SYSTEM support.

### PR DESCRIPTION
**Pull request by lrusak**
This allows us to add multiple devices in the future (such as H5 and A64)

to build a tarball for all H3 systems run
```
PROJECT=Allwinner ARCH=arm DEVICE=H3 make
```
to build an image with uboot for a specific board run

```
PROJECT=Allwinner ARCH=arm DEVICE=H3 UBOOT_SYSTEM=orangepi_pc make image
```

An image is needed for each board because each board has a different uboot config and dtb

to build all available images run something like
```
for board in $(./scripts/uboot_helper Allwinner H3); do
  PROJECT=Allwinner ARCH=arm DEVICE=H3 UBOOT_SYSTEM=${board} make image
done
```

This produces:

device specific images
```
LibreELEC-H3.arm-9.0-devel-20171018233138-r27174-gc66263b-orangepi_pc.img.gz
LibreELEC-H3.arm-9.0-devel-20171018233337-r27174-gc66263b-orangepi_lite.img.gz
LibreELEC-H3.arm-9.0-devel-20171018233522-r27174-gc66263b-orangepi_2.img.gz
LibreELEC-H3.arm-9.0-devel-20171018233643-r27174-gc66263b-orangepi_plus2e.img.gz
LibreELEC-H3.arm-9.0-devel-20171018233805-r27174-gc66263b-orangepi_plus.img.gz
LibreELEC-H3.arm-9.0-devel-20171018233928-r27174-gc66263b-orangepi_pc_plus.img.gz
LibreELEC-H3.arm-9.0-devel-20171018234049-r27174-gc66263b-orangepi_one.img.gz
```

project tar for all H3 systems
```
LibreELEC-H3.arm-9.0-devel-20171018234227-r27174-gc66263b.tar
```
----------

This PR moves all the heavy lifting to the new `scripts/uboot_helper` script. This script is designed to provide info for dtb, uboot config, uboot image, seek, blocksize and more. I think this is the best way forward rather than a bunch of `#ifdefs`

u-boot will be rebuilt each time UBOOT_SYSTEM is changed as a specific stamp for the u-boot package is created. if UBOOT_SYSTEM is the same u-boot will not be rebuilt.

This also doesn't install u-boot to the project tarball (if not defining UBOOT_SYSTEM) this is so we can have a tarball that can be used for all boards. u-boot will not be updated but the normal update will occur and the dtb files will be updated as well. If someone wants to update u-boot they just have to update using a newer device specific image.

Note that this breaks u-boot for Odroid_C2 and imx6 for now. I would like to get imx6 to use this method also and to move the Odroid_C2 specific u-boot to a separate package.

Please comment.

### CODE REVIEW

**Comment by vpeter4** [comment]
Not really important atm because there is no imx6 support in kodi 18 but using only mainline u-boot kills support for some solidrun's boards. As discussed on irc some patches are required.
If imx6 project will die this is not the issue then.

**Comment by jernejsk** [comment]
Most of the linux patches moved to device subdirectory would be also useful for other SoCs, but that can be decided when the support for new SoC is added.

Otherwise, PR looks fine by me.

**Comment by Kivutar** [comment]
I think it would be easier to just use:

```
PROJECT=Allwinner ARCH=arm DEVICE=orangepi_pc make image
PROJECT=Allwinner ARCH=arm DEVICE=cubietruck make image
PROJECT=Allwinner ARCH=arm DEVICE=cubieboard make image
PROJECT=Allwinner ARCH=arm DEVICE=orangepi_pc_plus make image

```

Because anyway, it's all sunxi, and they can use the same build dir, the same kernel and same rootfs. They can even use the same u-boot, with multiple defconfig backed in. The only difference will be the script.bin pointing to the right dtb.

**Comment by jernejsk** [comment]
Since we are planning to add support for A64 and H5 too which are similar to H3, but 64 bit, we have to have some differentiation. For example, H5 is exactly the same as H3, except Cortex A53 cores and mali450 GPU. Because of that, we can't have same build dir, different compiler flags, different (64 bit) kernel and different mali driver (there is no unified driver for mali4xx series).

BTW, here we are using mainline U-Boot and Linux, so we can fortunately forget everything about script.bin.

**Comment by Kivutar** [comment]
OK, I get it about A64 and H5. Makese sense.

And about script.bin, I was meaning boot.scr.

**Comment by lrusak** [comment]
@jernejsk rebased and updated for the new UBOOT_SYSTEM

**Comment by jernejsk** [comment]
I will test asap, but it will take few hours to do clean build.

File: projects/Allwinner/devices/H3/bootloader/install [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Comment (Kwiboo) [inline projects/Allwinner/devices/H3/bootloader/install]: this is not needed, this file is sourced not executed

File: projects/Allwinner/devices/H3/bootloader/mkimage [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (Kwiboo) [inline projects/Allwinner/devices/H3/bootloader/mkimage]: same as install, this is not needed, this file is sourced not executed

File: projects/Allwinner/devices/H3/bootloader/mkimage [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+if [ -f "$RELEASE_DIR/3rdparty/bootloader/u-boot.img" ]; then
+  echo "Writing u-boot.img to $(basename $DISK)"
+  dd if="$RELEASE_DIR/3rdparty/bootloader/u-boot.img" of="$DISK" bs=1K seek=40 conv=fsync,notrunc >"$SAVE_ERROR" 2>&1 || show_error
+fi
+
+if [ -f "$RELEASE_DIR/3rdparty/bootloader/SPL" ]; then
```
Comment (Kwiboo) [inline projects/Allwinner/devices/H3/bootloader/mkimage]: this checks for `SPL` but tries to write `sunxi-spl.bin`

File: projects/Allwinner/devices/H3/bootloader/update.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    fi
+  done
+
+# update bootloader files
+  if [ -f $SYSTEM_ROOT/usr/share/bootloader/u-boot.img ]; then
+    echo "*** updating u-boot image on: $BOOT_DISK ..."
+    dd if=$SYSTEM_ROOT/usr/share/bootloader/u-boot.img of="$BOOT_DISK" bs=1K seek=40 conv=fsync &>/dev/null
+  fi
+
+  if [ -f $SYSTEM_ROOT/usr/share/bootloader/SPL ]; then
```
Comment (Kwiboo) [inline projects/Allwinner/devices/H3/bootloader/update.sh]: same as mkimage, checks for `SPL` but tries to write `sunxi-spl.bin`

**Comment by lrusak** [comment]
@Kwiboo thanks, fixed

**Comment by jernejsk** [comment]
There is one thing I noticed before - there is 112 dtbs in image. How can we lower that to only one?

**Comment by jernejsk** [comment]
@lrusak KERNEL_MAKE_EXTRACMD="dtbs" causes compilation of all supported dtbs which is huge number. If this is removed, only sun8i-h3-orangepi-pc.dtb is built since it is the only one in KERNEL_UBOOT_EXTRA_TARGET.

But now that we have uboot_helper, can we automate this? Maintaining list of all supported dtbs on two places seems not necessary.

**Comment by vpeter4** [comment]
Also what is relationship with DEVICE and UBOOT_SYSTEM? I assume now every single image is exactly for one device? Consider i.MX6 devices from solidrun where only device tree's are different. Everything else is the same. It would be stupid to build different images for them.

**Comment by jernejsk** [comment]
@vpeter4 DEVICE=<soc> UBOOT_SYSTEM=<board> If you leave out UBOOT_SYSTEM, then generated update file will be universal for all board with same SoC.

**Comment by jernejsk** [comment]
@lrusak will you make any fixup or should I merge?

**Comment by lrusak** [comment]
I think I'm happy with this. It built fine for me with the uboot changes

**Pull request merged**